File: camera_rig/parser.py
"""
Parser for device calibration XML files.
"""

import numpy as np
from dataclasses import dataclass
from typing import List, Optional
from lxml import etree

from .transforms import parse_vector3, parse_rotation_matrix


@dataclass
class Camera:
    """Camera parameters."""
    name: str
    id: int
    model: str
    
    # Intrinsics
    width: int
    height: int
    focal_length: np.ndarray  # (fx, fy)
    principal_point: np.ndarray  # (cx, cy)
    radial_distortion: np.ndarray
    
    # Extrinsics (in rig coordinate system)
    rotation: np.ndarray  # 3x3 rotation matrix
    translation: np.ndarray  # 3D translation vector
    
    # Time alignment
    time_delta: float = 0.0


@dataclass
class IMU:
    """IMU parameters from SFConfig."""
    # IMU orientation relative to body frame (from ombc)
    rotation: np.ndarray  # 3x3 rotation matrix
    # IMU position relative to body frame (from tbc)
    translation: np.ndarray  # 3D translation vector


@dataclass
class VirtualDevice:
    """Virtual device for visualization."""
    name: str
    device_type: str  # "camera" or "axis_only"
    
    # Pose in world coordinates
    rotation: np.ndarray  # 3x3 rotation matrix
    translation: np.ndarray  # 3D translation vector
    
    # Camera parameters (only used if device_type == "camera")
    focal_length: Optional[np.ndarray] = None
    principal_point: Optional[np.ndarray] = None
    width: Optional[int] = None
    height: Optional[int] = None
    model: Optional[str] = None


@dataclass
class CameraRig:
    """Complete camera rig configuration."""
    device_uid: str
    cameras: List[Camera]
    imu: Optional[IMU]
    virtual_devices: List[VirtualDevice] = None
    
    def __post_init__(self):
        if self.virtual_devices is None:
            self.virtual_devices = []


def parse_calibration_file(xml_path: str) -> CameraRig:
    """
    Parse device calibration XML file.
    
    Args:
        xml_path: Path to device_calibration.xml
    
    Returns:
        CameraRig object with all parsed data
    """
    tree = etree.parse(xml_path)
    root = tree.getroot()
    
    device_uid = root.get('deviceUID', '')
    
    cameras = []
    for cam_elem in root.findall('Camera'):
        camera = parse_camera(cam_elem)
        cameras.append(camera)
    
    imu = None
    sfconfig = root.find('SFConfig')
    if sfconfig is not None:
        imu = parse_imu(sfconfig)
    
    virtual_devices = []
    virtual_root = root.find('VirtualDevices')
    if virtual_root is not None:
        for device_elem in virtual_root.findall('Device'):
            virtual_device = parse_virtual_device(device_elem)
            virtual_devices.append(virtual_device)
    
    reference_camera = None
    for cam in cameras:
        if cam.name == 'trackingA':
            reference_camera = cam
            break
    
    if reference_camera is None:
        raise ValueError("Reference camera 'trackingA' not found")
    
    R_ref = reference_camera.rotation
    t_ref = reference_camera.translation
    
    for camera in cameras:
        camera.rotation, camera.translation = transform_to_world(
            camera.rotation, camera.translation, R_ref, t_ref
        )
    
    for device in virtual_devices:
        pass
    
    return CameraRig(
        device_uid=device_uid,
        cameras=cameras,
        imu=imu,
        virtual_devices=virtual_devices
    )


def parse_camera(cam_elem) -> Camera:
    """Parse a single Camera element."""
    name = cam_elem.get('cam_name')
    cam_id = int(cam_elem.get('id'))
    
    calib = cam_elem.find('Calibration')
    model = calib.get('model')
    size = calib.get('size').strip().split()
    width, height = int(size[0]), int(size[1])
    
    focal = calib.get('focal_length').strip().split()
    focal_length = np.array([float(focal[0]), float(focal[1])])
    
    principal = calib.get('principal_point').strip().split()
    principal_point = np.array([float(principal[0]), float(principal[1])])
    
    radial = calib.get('radial_distortion').strip().split()
    radial_distortion = np.array([float(r) for r in radial])
    
    rig = cam_elem.find('Rig')
    translation = parse_vector3(rig.get('translation'))
    rotation = parse_rotation_matrix(rig.get('rowMajorRotationMat'))
    
    time_elem = cam_elem.find('TimeAlignment')
    time_delta = float(time_elem.get('delta')) if time_elem is not None else 0.0
    
    return Camera(
        name=name,
        id=cam_id,
        model=model,
        width=width,
        height=height,
        focal_length=focal_length,
        principal_point=principal_point,
        radial_distortion=radial_distortion,
        rotation=rotation,
        translation=translation,
        time_delta=time_delta
    )


def parse_imu(sfconfig_elem) -> IMU:
    """Parse IMU parameters from SFConfig."""
    state_init = sfconfig_elem.find('Stateinit')
    
    ombc_text = state_init.get('ombc')
    tbc_text = state_init.get('tbc')
    
    rotation = parse_rotation_matrix(ombc_text) if ombc_text else np.eye(3)
    translation = parse_vector3(tbc_text) if tbc_text else np.zeros(3)
    
    return IMU(rotation=rotation, translation=translation)


def parse_virtual_device(device_elem) -> VirtualDevice:
    """Parse a VirtualDevice element."""
    name = device_elem.get('name')
    device_type = device_elem.get('type', 'axis_only')
    
    translation_elem = device_elem.find('translation')
    rotation_elem = device_elem.find('rotation')
    
    translation = parse_vector3(translation_elem.text) if translation_elem is not None else np.zeros(3)
    
    if rotation_elem is not None:
        euler = rotation_elem.text.strip().split()
        roll, pitch, yaw = float(euler[0]), float(euler[1]), float(euler[2])
        from .transforms import euler_to_rotation_matrix
        rotation = euler_to_rotation_matrix(roll, pitch, yaw)
    else:
        rotation = np.eye(3)
    
    return VirtualDevice(
        name=name,
        device_type=device_type,
        rotation=rotation,
        translation=translation
    )


def transform_to_world(R_cam: np.ndarray, t_cam: np.ndarray,
                       R_ref: np.ndarray, t_ref: np.ndarray) -> tuple:
    """
    Transform camera pose to world coordinates using reference camera.
    
    The reference camera (trackingA) is at world origin with identity rotation.
    All other cameras are transformed relative to it.
    
    Args:
        R_cam: Camera rotation in rig coordinates
        t_cam: Camera translation in rig coordinates
        R_ref: Reference camera rotation (should be identity for trackingA)
        t_ref: Reference camera translation (should be zero for trackingA)
    
    Returns:
        (R_world, t_world) - camera pose in world coordinates
    """
    R_world = R_ref.T @ R_cam
    t_world = R_ref.T @ (t_cam - t_ref)
    return R_world, t_world

File: web/js/parser.js
/**
 * XML Parser for device_calibration.xml
 */

class CameraRigParser {
    constructor() {
        this.cameras = [];
        this.imu = null;
        this.virtualDevices = [];
        this.deviceUid = '';
    }

    parse(xmlString) {
        const parser = new DOMParser();
        const xmlDoc = parser.parseFromString(xmlString, 'text/xml');
        
        const root = xmlDoc.querySelector('DeviceConfiguration');
        if (!root) {
            throw new Error('Invalid XML: root element not found');
        }
        
        this.deviceUid = root.getAttribute('deviceUID') || '';
        
        this.parseCameras(root);
        this.parseIMU(root);
        this.parseVirtualDevices(root);
        
        return {
            deviceUid: this.deviceUid,
            cameras: this.cameras,
            imu: this.imu,
            virtualDevices: this.virtualDevices
        };
    }

    parseCameras(root) {
        const cameraElements = root.querySelectorAll('Camera');
        
        cameraElements.forEach(camElem => {
            const camera = {
                name: camElem.getAttribute('cam_name') || '',
                id: parseInt(camElem.getAttribute('id') || '0'),
                model: '',
                width: 0,
                height: 0,
                focalLength: [0, 0],
                principalPoint: [0, 0],
                radialDistortion: [],
                rotation: [[1,0,0],[0,1,0],[0,0,1]],
                translation: [0, 0, 0],
                timeDelta: 0
            };
            
            const calib = camElem.querySelector('Calibration');
            if (calib) {
                camera.model = calib.getAttribute('model') || 'PINHOLE';
                
                const size = (calib.getAttribute('size') || '640 480').trim().split(/\s+/);
                camera.width = parseInt(size[0]) || 640;
                camera.height = parseInt(size[1]) || 480;
                
                const focal = (calib.getAttribute('focal_length') || '0 0').trim().split(/\s+/);
                camera.focalLength = [parseFloat(focal[0]) || 0, parseFloat(focal[1]) || 0];
                
                const principal = (calib.getAttribute('principal_point') || '0 0').trim().split(/\s+/);
                camera.principalPoint = [parseFloat(principal[0]) || 0, parseFloat(principal[1]) || 0];
                
                const radial = (calib.getAttribute('radial_distortion') || '').trim().split(/\s+/);
                camera.radialDistortion = radial.map(v => parseFloat(v) || 0);
            }
            
            const rig = camElem.querySelector('Rig');
            if (rig) {
                const trans = (rig.getAttribute('translation') || '0 0 0').trim().split(/\s+/);
                camera.translation = trans.map(v => parseFloat(v) || 0);
                
                const rotMat = (rig.getAttribute('rowMajorRotationMat') || '1 0 0 0 1 0 0 0 1').trim().split(/\s+/);
                camera.rotation = this.parseRotationMatrix(rotMat);
            }
            
            const timeElem = camElem.querySelector('TimeAlignment');
            if (timeElem) {
                camera.timeDelta = parseFloat(timeElem.getAttribute('delta')) || 0;
            }
            
            this.cameras.push(camera);
        });
    }

    parseIMU(root) {
        const sfconfig = root.querySelector('SFConfig');
        if (!sfconfig) return;
        
        const stateInit = sfconfig.querySelector('Stateinit');
        if (!stateInit) return;
        
        const ombc = (stateInit.getAttribute('ombc') || '0 0 0').trim().split(/\s+/);
        const tbc = (stateInit.getAttribute('tbc') || '0 0 0').trim().split(/\s+/);
        
        this.imu = {
            rotation: this.parseRotationMatrix(ombc.map(v => parseFloat(v) || 0)),
            translation: tbc.map(v => parseFloat(v) || 0)
        };
    }

    parseVirtualDevices(root) {
        const virtualRoot = root.querySelector('VirtualDevices');
        if (!virtualRoot) return;
        
        const deviceElements = virtualRoot.querySelectorAll('Device');
        
        deviceElements.forEach(elem => {
            const device = {
                name: elem.getAttribute('name') || '',
                type: elem.getAttribute('type') || 'axis_only',
                translation: [0, 0, 0],
                rotation: [[1,0,0],[0,1,0],[0,0,1]]
            };
            
            const transElem = elem.querySelector('translation');
            if (transElem && transElem.textContent) {
                const trans = transElem.textContent.trim().split(/\s+/);
                device.translation = trans.map(v => parseFloat(v) || 0);
            }
            
            const rotElem = elem.querySelector('rotation');
            if (rotElem && rotElem.textContent) {
                const euler = rotElem.textContent.trim().split(/\s+/).map(v => parseFloat(v) || 0);
                device.rotation = this.eulerToRotationMatrix(euler[0], euler[1], euler[2]);
            }
            
            const focalElem = elem.querySelector('focalLength');
            if (focalElem && focalElem.textContent) {
                const focal = focalElem.textContent.trim().split(/\s+/);
                device.focalLength = [parseFloat(focal[0]) || 200, parseFloat(focal[1]) || 200];
            }
            
            const principalElem = elem.querySelector('principalPoint');
            if (principalElem && principalElem.textContent) {
                const principal = principalElem.textContent.trim().split(/\s+/);
                device.principalPoint = [parseFloat(principal[0]) || 320, parseFloat(principal[1]) || 240];
            }
            
            const resolution = elem.querySelector('resolution');
            if (resolution) {
                device.width = parseInt(resolution.getAttribute('width')) || 640;
                device.height = parseInt(resolution.getAttribute('height')) || 480;
            }
            
            this.virtualDevices.push(device);
        });
    }

    parseRotationMatrix(values) {
        if (values.length === 3) {
            const axis = [values[0], values[1], values[2]];
            const angle = Math.sqrt(axis[0]**2 + axis[1]**2 + axis[2]**2);
            if (angle < 1e-10) {
                return [[1,0,0],[0,1,0],[0,0,1]];
            }
            return this.axisAngleToRotationMatrix(
                [axis[0]/angle, axis[1]/angle, axis[2]/angle],
                angle
            );
        }
        
        if (values.length === 9) {
            return [
                [values[0], values[1], values[2]],
                [values[3], values[4], values[5]],
                [values[6], values[7], values[8]]
            ];
        }
        
        return [[1,0,0],[0,1,0],[0,0,1]];
    }

    axisAngleToRotationMatrix(axis, angle) {
        const c = Math.cos(angle);
        const s = Math.sin(angle);
        const t = 1 - c;
        const [x, y, z] = axis;
        
        return [
            [t*x*x + c,    t*x*y - z*s,  t*x*z + y*s],
            [t*x*y + z*s,  t*y*y + c,    t*y*z - x*s],
            [t*x*z - y*s,  t*y*z + x*s,  t*z*z + c  ]
        ];
    }

    eulerToRotationMatrix(roll, pitch, yaw) {
        const r = roll * Math.PI / 180;
        const p = pitch * Math.PI / 180;
        const y = yaw * Math.PI / 180;
        
        const Rx = [
            [1, 0, 0],
            [0, Math.cos(r), -Math.sin(r)],
            [0, Math.sin(r), Math.cos(r)]
        ];
        
        const Ry = [
            [Math.cos(p), 0, Math.sin(p)],
            [0, 1, 0],
            [-Math.sin(p), 0, Math.cos(p)]
        ];
        
        const Rz = [
            [Math.cos(y), -Math.sin(y), 0],
            [Math.sin(y), Math.cos(y), 0],
            [0, 0, 1]
        ];
        
        const RyRx = this.multiplyMatrix(Ry, Rx);
        return this.multiplyMatrix(Rz, RyRx);
    }

    multiplyMatrix(A, B) {
        const result = [[0,0,0],[0,0,0],[0,0,0]];
        for (let i = 0; i < 3; i++) {
            for (let j = 0; j < 3; j++) {
                for (let k = 0; k < 3; k++) {
                    result[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        return result;
    }
}

window.CameraRigParser = CameraRigParser;
